proposals/0013-expanded-comparison-sampling.md

```diff
@@ -0,0 +1,441 @@
+ difficult or impossible.
+
+## Proposed solution
+
+To add full comparison sampling support, HLSL will introduce an optional feature
+ that adds these new comparison sampling builtin methods to all texture objects
+ that currently support the non-comparison methods:
+
+* `SampleCmpBias(SamplerComparisonState S, float<crds> Location, float Bias)`
+  performs a comparison sample using the implicitly-determined LOD with the
+  explicitly-provided `Bias` value added to that result.
+* `SampleCmpGrad(SamplerState S, float<crds> Location, float DDX, float DDY)`
+   samples using the LOD calculated using the explicitly-provided
+   `DDX` and `DDY` gradients.
+
+Only the simplest overloads of the new and existing builtin methods are
+ described in detail here for the sake of brevity,
+Though not included above, all relevant overloads of `SampleBias` and
+ `SampleGrad` will be reflected in their new `SampleCmpBias` and
+ `SampleCmpGrad` counterparts.
+The parameters not explicitly described here will function exactly as they do
+ when used by the non-comparison builtin methods:
+
+* `SampleCmpBias(SamplerComparisonState S, float<crds> Location, float Bias)`
+* `SampleCmpBias(SamplerComparisonState S, float<crds> Location, float Bias, int<dims> Offset)`
+* `SampleCmpBias(SamplerComparisonState S, float<crds> Location, float Bias, int<dims> Offset, float Clamp)`
+* `SampleCmpBias(SamplerComparisonState S, float<crds> Location, float Bias, int<dims> Offset, float Clamp, uint Status)`
+%* `SampleCmpGrad(SamplerState S, float<crds> Location, float<dims> DDX, float<dims> DDY)`
+* `SampleCmpGrad(SamplerState S, float<crds> Location, float<dims> DDX, float<dims> DDY, int<dims> Offset)`
+* `SampleCmpGrad(SamplerState S, float<crds> Location, float<dims> DDX, float<dims> DDY, int<dims> Offset, float Clamp)`
+* `SampleCmpGrad(SamplerState S, float<crds> Location, float<dims> DDX, float<dims> DDY, int<dims> Offset, float Clamp, uint Status)`
+
+To add LOD calculation for comparison samplers, HLSL will introduce and require
+ these new comparison sampler method overloads to all texture objects that
+ currently support the non-comparison method overloads:
+
+* `CalculateLevelOfDetailUnclamped(SamplerComparisonState S, float<crds> Location)`
+  returns the raw LOD value determined implicitly using gradients representing
+  how much the `Location` changes across neighboring pixels regardless of
+  whether it falls outside the valid range.
+* `CalculateLevelOfDetail(SamplerComparisonState S, float<crds> Location)`
+  returns the implicitly-determined LOD value clamped to the valid range.
+
+## Detailed Design
+
+### HLSL additions
+
+These new HLSL texture methods will be added:
+
+```c++
+template<typename ElementType, int CoordsCt, int DimsCt>
+vector<ElementType, ElementSize> TextureBase::SampleCmpGrad(
+    SamplerComparisonState S,
+    vector<float,CoordsCt> Location,
+    vector<float,DimsCt> DDX,
```

Comment:
Why switch from using `crds` and `dims` above to `CoordsCt` and `DimsCt` here?  The other ones had a description for what they meant at least.

Reply:
I bet it was due to search pattern being broken, where the `float<crds>` form is used elsewhere, instead of `vector<float,CoordsCt>`.

Reply:
It was to try to make things clearer. Since we're using templates, I felt somehow I could be more verbose without being confusing. I should define them anyway, but some of this needs to be spread to other locations too, I'm just waiting to see if this addresses Chris' comments before I apply it elsewhere.

Reply:
I've since propagated all the template usage everywhere